fix: API gateway JWT documentation for EdgeX 3.0 (#1029)
edgexfoundry/edgex-docs: docs_src/security/Ch-APIGateway.md

### Configuration of Adding Microservices Routes for API Gateway

A standard set of routes are configured statically via the `security-proxy-setup` microservice.
Additional routes can be added via the `EDGEX_ADD_PROXY_ROUTE` environment variable.
Here is an example:

```yaml
security-proxy-setup:
      ...
    environment:
      ...
      EDGEX_ADD_PROXY_ROUTE: "app-myservice.http://edgex-app-myservice:56789"
      ...

...

app-myservice:
   ...
   container_name: app-myservice-container
   hostname: edgex-app-myservice
   ...
  
```

The value of `EDGEX_ADD_PROXY_ROUTE` takes a comma-separated list
of one or more paired additional prefix and URL
for which to create proxy routes.
The paired specification is given as the following:

    <RoutePrefix>.<TargetRouteURL>

where **RoutePrefix** is the base path that will be created off of the root
of the API gateway to route traffic to the target.
This should typically be the service key that the app uses to register
with the EdgeX secret store and configuration provider,
as the name of the service in the docker-compose file has
security implications when using delayed-start services.

> **TargetRouteURL** is the full qualified URL for the target service,

Reviewer: ```suggestion
**TargetRouteURL** is the fully qualified URL for the target service,
```

Author: Fixed in https://github.com/bnevis-i/edgex-docs/commit/2fff06e6ba4522a0ffa6a91f3b1dc29a138b3c95

  
```

The value of `EDGEX_ADD_PROXY_ROUTE` takes a comma-separated list

> of one or more (at least one) paired additional service name and URL

Reviewer: ```suggestion
of one or more paired additional prefix and URL
```

for which to create proxy routes.
The paired specification is given as the following:

    <RoutePrefix>.<TargetRouteURL>

> where **RoutePrefix** is a directory path that will be created off of the root
> of the API gateway to route traffic to the target.

Reviewer: ```suggestion
where **RoutePrefix** is the base path that will be created off of the root
of the API gateway to route traffic to the target. This should typically be the service name.
```

**TargetRouteURL** is the full qualified URL for the target service,
like `http://myapp:56789` as it is known on on the network on which the API gateway is running.
For Docker, the hostname should match the hostname specified in the `docker-compose.yml` file.

So as an example, for a single service, the value of `EDGEX_ADD_PROXY_ROUTE` would be:  "`myApp.http://myapp:56789`".

In the given example,
```
EDGEX_ADD_PROXY_ROUTE: "myApp.http://my-app:56789"
```

When a request to the API gateway is received,
such as `GET https://localhost:8443/myApp/api/v2/ping`,
the API gateway will reissue the request as

> `GET http://myapp:56789/api/v2/ping`.

Reviewer: Isn't `my-app-service` same as `my-app?

It is hard to follow because these are all similar names. It is also easy to make a mistake while reading or editing the text. I suggest embedding the label inside the example and keeping a small set of them: `service-hostname` (instead of my-app), `service-name` (instead of myApp).

> `GET http://myapp:56789/api/v2/ping`.

Reviewer: I think it would be least confusing if you stick with a a single naming convention for both route base name (service name) and hostname.

> `GET http://myapp:56789/api/v2/ping`.

Author: > Isn't `my-app-service` same as `my-app?
> 
> It is hard to follow because these are all similar names just presented in different naming conventions. It is also easy to make a mistake while reading or editing the text. I suggest embedding the label inside the example and keeping a small set of them: `service-hostname` (instead of my-app), `service-name` (instead of myApp).

This comment was made in flight as I was addressing the previous comments.  Diregarding for now. Please see latest commit.

    curl http://<core-data-microservice-ip>:59880/api/v2/ping

> Where (\<core-data-microservice-ip\> is the Docker IP address of

Reviewer: Not introduced by this PR, but I don't know why `<>` are being escaped.
```suggestion
Where (`<core-data-microservice-ip>` is the Docker IP address of
```

Author: I tend to leave markdown syntax I don't understand alone.... but changed anyway.

Author: Fixed in https://github.com/bnevis-i/edgex-docs/commit/2fff06e6ba4522a0ffa6a91f3b1dc29a138b3c95

### Creating Access Token for API Gateway Authentication

Authentication is more fully explained

> in the [authentication chapter](./Ch-Authenticating.md).

Reviewer: Note that the EdgeX UI's login page gives a link to this chapter, hardcoded for v2.1: https://docs.edgexfoundry.org/2.1/security/Ch-APIGateway/#creating-access-token-for-api-gateway-authentication

That should perhaps be replaced with https://docs.edgexfoundry.org/3.0/security/Ch-Authenticating/

Author: https://github.com/edgexfoundry/edgex-ui-go/pull/598

place, here is the sample request to hit the ping endpoint of the EdgeX
Core Data microservice using curl:

    curl http://<core-data-microservice-ip>:59880/api/v2/ping

With the security service and JWT authentication is enabled, the command
changes to:

    curl -k -H 'Authorization: Bearer <JWT>' https://myhostname:8443/core-data/api/v2/ping

In summary the difference between the two commands are listed below:

-   `-k` tells curl to ignore certificate errors. This is for
    demonstration purposes. In production, a known certificate that
    the client trusts be installed on the proxy and this parameter omitted.
-   Use the https versus http protocol identifier for SSL/TLS secure
    communication.
-   The service port 8443 is the default TLS service port of API gateway
-   Use the URL path "core-data" to indicate which EdgeX microservice
    the request is routed to

> -   Use header of `-H "Authorization: Bearer \<JWT\>"` to

Reviewer: ```suggestion
-   Use header of `-H "Authorization: Bearer <JWT>"` to
```

Author: Fixed in https://github.com/bnevis-i/edgex-docs/commit/2fff06e6ba4522a0ffa6a91f3b1dc29a138b3c95
